### ISSUE TITLE
EFS restore segfault

### ISSUE THREAD

**Issue by grke** [grke/burp#334]
Report from the mailing list:

> Program received signal SIGSEGV, Segmentation fault.
> 0x000007fefeb5a997 in CloseServiceHandle ()
> from C:\Windows\system32\advapi32.dll
> (gdb) bt
> #0 0x000007fefeb5a997 in CloseServiceHandle ()
> from C:\Windows\system32\advapi32.dll
> #1 0x00000000004187e1 in transfer_efs_in (sb=Unhandled dwarf expression
> opcode
> 0xf3
> )
> at /home/graham/burp-1.4.40/src/msg.c:176
> #2 transfer_gzfile_in (sb=Unhandled dwarf expression opcode 0xf3
> ) at /home/graham/burp-1.4.40/src/msg.c:205
> #3 0x0000000000419c9a in restore_file_or_get_meta (bfd=0x22ebc0,
> sb=0x22ec60,
> fname=0x285b2f0 "E:/RESTORE4/C_/Users/my_user/Folder1/Folder 2
> /Projects/SOMETHING-ELSE 2015-2018/Posters SOMEONE/__MACOSX/._poser
> 2.pdf",
> act=Unhandled dwarf expression opcode 0xf3
> ) at /home/graham/burp-1.4.40/src/restore_client.c:275
> #4 0x000000000041a3eb in do_restore_client (conf=0x22f9c0,
> act=ACTION_RESTORE, vss_restore=1, p1cntr=0x22f6a0, cntr=0x22f490)
> at /home/graham/burp-1.4.40/src/restore_client.c:833
> #5 0x000000000040735b in do_client (conf=0x22f9c0, act=ACTION_RESTORE,
> vss_restore=1, json=0) at /home/graham/burp-1.4.40/src/client.c:573
> #6 0x00000000004074e5 in client (conf=0x22f9c0, act=ACTION_RESTORE,
> vss_restore=1, json=0) at /home/graham/burp-1.4.40/src/client.c:635
> #7 0x0000000000419638 in BurpMain (argc=Unhandled dwarf expression opcode
> 0xf3
> )
> at /home/graham/burp-1.4.40/src/prog.c:373
> #8 0x000000000044a411 in main (argc=9, argv=0xb04870)
> at ../libwin32/main.cpp:39
> (gdb)

**Comment by grke** [comment]
I am not sure right now if EFS backup/restores are completely broken in
burp, or if this is a problem with trying to restore an EFS file back to a
different machine than it came from.
The solution for you might be to restore the EFS keys first, and then the
EFS files. But I do not know where the EFS keys live, or if they get backed
up with VSS. Maybe a Windows expert on the list would know that kind of
thing.
Obviously, burp should not segfault, but it might be some time before I get
round to setting up an EFS test and trying to reproduce it.

**Comment by grke** [comment]
I just tried on Windows 7 32bit, and EFS restores back to the original machine aren't working for me, although there is no segfault. I tried 1.4.40 and 2.0.x.

**Comment by grke** [comment]
So, I found an issue in 2.0.26, where the server wasn't sending the correct commands on an EFS restore. I have fixed that and it is now in master.

I started on 1.4.40, both sides, and made an EFS backup.
The 1.4.40 restore failed.
I upgraded the server to 2.0.26+fix, and the restore failed.
I upgraded the client to 2.0.26, and the restore succeeded, meaning the backup from 1.4.40 was good.
I then tried to restore to a different Windows machine.
The restore succeeded, and I could see all the files on Windows, but Windows was unable to open them, as you'd expect, as it didn't have the EFS keys.

I think the segfault will go away once you upgrade server and client from 1.4.40 to 2.0.26+fix.
The fix will be in the next release. I will leave this issue open until I include at least one EFS file in
the automatic burp tests.
